File: docs/examples/x-wallet-dev/frontend/src/api/sponsor.ts
import { API_BASE_URL } from '../utils/constants';
import type {
  SponsorTxRequestBody,
  CreateSponsoredTransactionApiResponse,
  ExecuteSponsoredTransactionApiInput,
  ExecuteSponsoredTransactionApiResponse,
} from '../types';

/**
 * Create a sponsored transaction using Enoki via backend
 *
 * @param body - The sponsor request containing network, txBytes, sender, and optional allowedAddresses
 * @returns The sponsored transaction bytes and digest
 */
export async function createSponsoredTransaction(
  body: SponsorTxRequestBody
): Promise<CreateSponsoredTransactionApiResponse> {
  const response = await fetch(`${API_BASE_URL}/api/sponsor`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      network: body.network,
      txBytes: body.txBytes,
      sender: body.sender,
      allowedAddresses: body.allowedAddresses || [],
    }),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.error || 'Failed to create sponsored transaction');
  }

  return await response.json();
}

/**
 * Execute a sponsored transaction using Enoki via backend
 *
 * @param body - The execute request containing digest and signature
 * @returns The final transaction digest
 */
export async function executeSponsoredTransaction(
  body: ExecuteSponsoredTransactionApiInput
): Promise<ExecuteSponsoredTransactionApiResponse> {
  const response = await fetch(`${API_BASE_URL}/api/execute`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.error || 'Failed to execute sponsored transaction');
  }

  return await response.json();
}

File: docs/examples/x-wallet-dev/backend/src/api/sponsor.rs
//! Transaction sponsorship API handlers (Enoki)
//!
//! - sponsor_transaction
//! - execute_sponsored_transaction

use axum::{extract::State, http::StatusCode, Json};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

use crate::clients::enoki::EnokiClient;
use crate::webhook::handler::AppState;

// ====== Sponsor Transaction ======

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SponsorTxRequest {
    pub network: String,
    pub tx_bytes: String,
    pub sender: String,
    #[serde(default)]
    #[allow(dead_code)]
    pub allowed_addresses: Vec<String>,
}

#[derive(Debug, Serialize)]
pub struct SponsorTxResponse {
    pub bytes: String,
    pub digest: String,
}

#[derive(Debug, Serialize)]
pub struct SponsorErrorResponse {
    pub error: String,
}

/// Create a sponsored transaction using Enoki
pub async fn sponsor_transaction(
    State(state): State<Arc<AppState>>,
    Json(request): Json<SponsorTxRequest>,
) -> Result<Json<SponsorTxResponse>, (StatusCode, Json<SponsorErrorResponse>)> {
    tracing::info!(
        sender = %request.sender,
        network = %request.network,
        "Sponsor transaction request received"
    );

    let enoki_client = EnokiClient::new(
        state.config.enoki_api_key.clone(),
        request.network.clone(),
    );

    match enoki_client
        .create_sponsored_transaction(request.tx_bytes, request.sender)
        .await
    {
        Ok(response) => {
            tracing::info!(digest = %response.digest, "Transaction sponsored successfully");
            Ok(Json(SponsorTxResponse {
                bytes: response.bytes,
                digest: response.digest,
            }))
        }
        Err(err) => {
            tracing::error!("Failed to sponsor transaction: {:?}", err);
            Err((
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(SponsorErrorResponse {
                    error: format!("Could not create sponsored transaction: {}", err),
                }),
            ))
        }
    }
}

// ====== Execute Sponsored Transaction ======

#[derive(Debug, Deserialize)]
pub struct ExecuteSponsoredTxRequest {
    pub digest: String,
    pub signature: String,
}

#[derive(Debug, Serialize)]
pub struct ExecuteSponsoredTxResponse {
    pub digest: String,
}

/// Execute a sponsored transaction using Enoki
pub async fn execute_sponsored_transaction(
    State(state): State<Arc<AppState>>,
    Json(request): Json<ExecuteSponsoredTxRequest>,
) -> Result<Json<ExecuteSponsoredTxResponse>, (StatusCode, Json<SponsorErrorResponse>)> {
    tracing::info!(
        digest = %request.digest,
        "Execute sponsored transaction request received"
    );

    let enoki_client = EnokiClient::new(
        state.config.enoki_api_key.clone(),
        state.config.enoki_network.clone(),
    );

    match enoki_client
        .execute_sponsored_transaction(request.digest, request.signature)
        .await
    {
        Ok(response) => {
            tracing::info!(digest = %response.digest, "Sponsored transaction executed successfully");
            Ok(Json(ExecuteSponsoredTxResponse {
                digest: response.digest,
            }))
        }
        Err(err) => {
            tracing::error!("Failed to execute sponsored transaction: {:?}", err);
            Err((
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(SponsorErrorResponse {
                    error: format!("Could not execute sponsored transaction: {}", err),
                }),
            ))
        }
    }
}
